### PR TITLE
cleanning the tree_map code structure

### PR DESCRIPTION
**Pull request by MoustaphaSaad**

### CODE REVIEW

File: cpprelude/inc/cpprelude/tree_map.h [hunk 1]
```diff
@@ -0,0 +1,1065 @@
...
+			if (y == nullptr)
+				_root = z;
+			else
+			{
+				if (_less_than(z->data, y->data))
+					y->left = z;
+				else
+					y->right = z;
+			}
+			z->left = nullptr;
```
Comment (NoraRo) [inline cpprelude/inc/cpprelude/tree_map.h]: z->left = nullptr;
z->right = nullptr;
z->color = color_type::RED;
suppose thes lines are initialized in node_type constructor. So, we can remove them.

File: cpprelude/inc/cpprelude/iterator.h [hunk 2]
```diff
@@ -91,36 +91,24 @@ namespace cpprelude
...
+			red_black_tree_node()
+				:left(nullptr),
+				 right(nullptr),
+				 parent(nullptr),
+				 color(RED)
 			{}
 
-			rb_node(const T& d, color_type c = RED)
-				:data(d), color(c), left(nullptr), right(nullptr), parent(nullptr)
+			red_black_tree_node(const T& data_, color_type color_ = color_type::RED)
```
Comment (NoraRo) [inline cpprelude/inc/cpprelude/iterator.h]: yep, we should initialize left, right, parent to nullptr.

File: cpprelude/inc/cpprelude/tree_map.h [hunk 3]
```diff
@@ -0,0 +1,1065 @@
...
+
+			it->data.~data_type();
+			_allocator.free(make_slice(it));
+			--_count;
+		}
+
+		void
+		_reset(node_type* it)
+		{
+			if(it == nullptr)
```
Comment (NoraRo) [inline cpprelude/inc/cpprelude/tree_map.h]: we can change this to use preorder_traverse function like _copy_content.

Comment (NoraRo) [inline cpprelude/inc/cpprelude/tree_map.h]: *postorder_traverse

File: scratch/src/benchmark.cpp [hunk 4]
```diff
@@ -1635,7 +1639,11 @@ benchmark_std_map(cpprelude::usize limit)
 
 		w.start();
 		for (cpprelude::usize i = 0; i < limit; ++i)
-			tree.insert(std::make_pair(i, true));
+		{
+			tree.insert({i, true});
+			auto it = tree.find(i);
```
Comment (NoraRo) [inline scratch/src/benchmark.cpp]: find and erase when only one element in the tree will not test the speed. can we change it to 2 sperate loops one for insertion and one for find and erase?

File: cpprelude/inc/cpprelude/tree_map.h (hunk 1 above)
Comment (MoustaphaSaad) [inline cpprelude/inc/cpprelude/tree_map.h]: done

File: cpprelude/inc/cpprelude/iterator.h (hunk 2 above)
Comment (MoustaphaSaad) [inline cpprelude/inc/cpprelude/iterator.h]: done

File: cpprelude/inc/cpprelude/tree_map.h (hunk 3 above)
Comment (MoustaphaSaad) [inline cpprelude/inc/cpprelude/tree_map.h]: nope this is better and i will tell you why

File: scratch/src/benchmark.cpp (hunk 4 above)
Comment (MoustaphaSaad) [inline scratch/src/benchmark.cpp]: :+1:

**Pull request merged**